### PR TITLE
Allow `ControllerDeployment` to reference chart from an OCI Repository

### PR DESCRIPTION
**Pull request by maboehm**
**How to categorize this PR?**
/area control-plane
/kind enhancement

**What this PR does / why we need it**:
Introduces a new field `.helm.ociRepository` which instructs the `controllerinstallation-reconciler` to pull the chart from a registry. An in-memory cache is used to avoid unnecessary requests (cache key is registry@digest). The local setup / skaffold now uses this for the provider local.

**Which issue(s) this PR fixes**:
https://github.com/gardener-community/hackathon/tree/main/2024-05_Schelklingen#%EF%B8%8F-oci-helm-release-reference-for-controllerdeployments

Part of https://github.com/gardener/gardener/issues/9635

**Special notes for your reviewer**:
This introduces 2 new direct dependencies - `github.com/google/go-containerregistry` for working with the registry and `github.com/distribution/distribution/v3` for starting a local registry during tests. Not sure if this is a problem, just wanted to mention it.

**Release note**:

```feature operator
It is now possible to specify an OCI repository in `ControllerDeployment`s describing from where the Helm chart can be pulled (instead of specifying a `base64`-encoded chart in the specification).
```

### CODE REVIEW

File: pkg/apis/core/v1/types_controllerdeployment.go [hunk 1]
```diff
@@ -5,6 +5,8 @@
 package v1
 
 import (
+	strings "strings"
```
Comment (rfranzke) [inline pkg/apis/core/v1/types_controllerdeployment.go]:
```suggestion
	"strings"
```

File: pkg/apis/core/v1/types_controllerdeployment.go [hunk 2]
```diff
@@ -43,4 +45,41 @@ type HelmControllerDeployment struct {
...
+	// Repository is a reference to an OCI artifact repository.
+	// +optional
+	Repository string `json:"repository,omitempty" protobuf:"bytes,2,name=repository"`
+	// Tag is the image tag to pull.
+	// +optional
+	Tag string `json:"tag,omitempty" protobuf:"bytes,3,opt,name=tag"`
+	// Digest of the image to pull, takes precedence over tag.
+	// The value should be in the format 'sha256:<HASH>'.
+	// +optional
+	Digest string `json:"digest,omitempty" protobuf:"bytes,4,opt,name=digest"`
```
Comment (rfranzke) [inline pkg/apis/core/v1/types_controllerdeployment.go]: *string

optional fields are usually pointer types

File: pkg/apis/core/v1/types_controllerdeployment.go [hunk 3]
```diff
@@ -43,4 +45,41 @@ type HelmControllerDeployment struct {
...
+		ref = r.Ref
+	case r.Digest != "":
+		// when digest is set we ignore the tag
+		ref = r.Repository + "@" + r.Digest
+	case r.Tag != "":
+		ref = r.Repository + ":" + r.Tag
+	}
+
+	ref = strings.TrimPrefix(ref, "oci://")
+	return ref
```
Comment (rfranzke) [inline pkg/apis/core/v1/types_controllerdeployment.go]: Nit:

```suggestion
	return strings.TrimPrefix(ref, "oci://")
```

File: pkg/apis/core/v1/types_controllerdeployment.go [hunk 4]
```diff
@@ -43,4 +45,41 @@ type HelmControllerDeployment struct {
...
+	Tag string `json:"tag,omitempty" protobuf:"bytes,3,opt,name=tag"`
+	// Digest of the image to pull, takes precedence over tag.
+	// The value should be in the format 'sha256:<HASH>'.
+	// +optional
+	Digest string `json:"digest,omitempty" protobuf:"bytes,4,opt,name=digest"`
+}
+
+// GetURL returns the fully-qualified OCIRepository URL of the artifact.
+func (r *OCIRepository) GetURL() string {
+	ref := ""
```
Comment (rfranzke) [inline pkg/apis/core/v1/types_controllerdeployment.go]: Nit:

```suggestion
	var ref string

```

File: pkg/apis/core/v1beta1/types_controllerdeployment.go [hunk 5]
```diff
@@ -53,4 +53,18 @@ type HelmControllerDeployment struct {
...
+// OCIRepository configures where to pull an OCI Artifact, that could contain for example a Helm Chart.
+type OCIRepository struct {
+	// Ref is the full artifact Ref and takes precedence over all other fields.
+	Ref string `json:"ref,omitempty" protobuf:"bytes,1,name=ref"`
+	// Repository is a reference to an OCI artifact repository.
+	Repository string `json:"repository,omitempty" protobuf:"bytes,2,name=repository"`
+	// Tag is the image tag to pull.
+	Tag string `json:"tag,omitempty" protobuf:"bytes,3,opt,name=tag"`
+	// Digest of the image to pull, takes precedence over tag.
+	Digest string `json:"digest,omitempty" protobuf:"bytes,4,opt,name=digest"`
```
Comment (rfranzke) [inline pkg/apis/core/v1beta1/types_controllerdeployment.go]: same here (*string)

also, `// +optional` tags are missing

File: pkg/apis/core/v1beta1/types_controllerdeployment.go [hunk 6]
```diff
@@ -53,4 +53,18 @@ type HelmControllerDeployment struct {
 	Chart []byte `json:"chart,omitempty" protobuf:"bytes,1,opt,name=chart"`
 	// Values is a map of values for the given chart.
 	Values *apiextensionsv1.JSON `json:"values,omitempty" protobuf:"bytes,2,opt,name=values"`
+	// OCIRepository defines where to pull the chart.
```
Comment (rfranzke) [inline pkg/apis/core/v1beta1/types_controllerdeployment.go:56]: // +optional

File: pkg/apis/core/types_controllerdeployment.go [hunk 7]
```diff
@@ -52,4 +52,19 @@ type HelmControllerDeployment struct {
...
+type OCIRepository struct {
+	// Ref is the full artifact Ref and takes precedence over all other fields.
+	Ref string
+	// Repository is a reference to an OCI artifact repository.
+	Repository string
+	// Tag is the image tag to pull.
+	Tag string
+	// Digest of the image to pull, takes precedence over tag.
+	// The value should be in the format 'sha256:<HASH>'.
+	Digest string
```
Comment (rfranzke) [inline pkg/apis/core/types_controllerdeployment.go]: *string

File: pkg/apis/core/validation/controllerdeployment.go [hunk 8]
```diff
@@ -58,9 +59,48 @@ func ValidateControllerDeploymentUpdate(new, _ *core.ControllerDeployment) field
 func validateHelmControllerDeployment(helmControllerDeployment *core.HelmControllerDeployment, fldPath *field.Path) field.ErrorList {
 	allErrs := field.ErrorList{}
 
-	// rawChart is required as long as there is no other source (OCI sources will be added later on)
-	if len(helmControllerDeployment.RawChart) == 0 {
-		allErrs = append(allErrs, field.Required(fldPath.Child("rawChart"), ""))
+	if (len(helmControllerDeployment.RawChart) == 0) == (helmControllerDeployment.OCIRepository == nil) {
+		allErrs = append(allErrs, field.Required(fldPath, "must provide either rawChart or ociRepository"))
+	}
```
Comment (rfranzke) [inline pkg/apis/core/validation/controllerdeployment.go:66]: I guess it makes sense to forbid specifying both at the same time?

File: pkg/apis/core/validation/controllerdeployment.go [hunk 9]
```diff
@@ -58,9 +59,48 @@ func ValidateControllerDeploymentUpdate(new, _ *core.ControllerDeployment) field
...
+
+	if oci.Ref != "" {
+		// all other fields must be empty if ref is set.
+		for name, val := range map[string]string{
+			"repository": oci.Repository,
+			"tag":        oci.Tag,
+			"digest":     oci.Digest,
+		} {
+			if val != "" {
+				allErrs = append(allErrs, field.Invalid(fldPath.Child(name), val, fmt.Sprintf("cannot provide %s when ref is set", name)))
```
Comment (rfranzke) [inline pkg/apis/core/validation/controllerdeployment.go]: I think `field.Forbidden` is more appropriate in this case?

File: pkg/apis/core/validation/controllerdeployment_test.go [hunk 10]
```diff
@@ -79,7 +79,20 @@ var _ = Describe("#ValidateControllerDeployment", func() {
 		}))))
 	})
 
-	Context("helm type", func() {
+	It("helm type: should not allow both rawChart and OCIRepository", func() {
```
Comment (rfranzke) [inline pkg/apis/core/validation/controllerdeployment_test.go]: Why did you remove the `	Context("helm type", func() {`? I think all these `It`() belong into it, this way the "helm type: "` can be avoided in their descriptions?

File: pkg/utils/oci/cache.go [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+// SPDX-License-Identifier: Apache-2.0
+
+package oci
+
+import "sync"
+
+var defaultCache = newCache()
+
+type cacher interface {
+	Get(k string) ([]byte, bool)
```
Comment (rfranzke) [inline pkg/utils/oci/cache.go]: Nit: Can we rename `k` to `key` to make it more clear? :)

File: pkg/utils/oci/fake/fake_registry.go [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+}
+
+// NewRegistry returns a new registry
+func NewRegistry() *Registry {
+	return &Registry{
+		artifacts: make(map[string][]byte),
+	}
+}
+
+// Pull implements registry.Interface
```
Comment (rfranzke) [inline pkg/utils/oci/fake/fake_registry.go]: `oci.Interface`, or?

File: pkg/utils/oci/helmregistry.go [hunk 13]
```diff
@@ -0,0 +1,153 @@
...
+	}
+
+	var digest gcrv1.Hash
+	desc, err := remote.Head(ref, opts...)
+	if err == nil {
+		digest = desc.Digest
+	} else {
+		rd, err := remote.Get(ref, opts...)
+		if err != nil {
+			return ""
```
Comment (rfranzke) [inline pkg/utils/oci/helmregistry.go]: Should we really "silently" ignore errors here?

File: pkg/utils/oci/helmregistry.go [hunk 14]
```diff
@@ -0,0 +1,153 @@
...
+	if err != nil {
+		return nil, err
+	}
+	remoteOpts := []remote.Option{
+		remote.WithContext(ctx),
+	}
+
+	if key := cacheKeyFromRef(ref, remoteOpts...); key != "" {
+		blob, found := r.cache.Get(key)
+		if found {
```
Comment (rfranzke) [inline pkg/utils/oci/helmregistry.go]: Nit: Could be put in one line

File: pkg/utils/oci/helmregistry.go [hunk 15]
```diff
@@ -0,0 +1,153 @@
...
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse layers: %w", err)
+	}
+
+	if len(layers) < 1 {
+		return nil, fmt.Errorf("no layers found")
+	}
+
+	var layer gcrv1.Layer
+	found := false
```
Comment (rfranzke) [inline pkg/utils/oci/helmregistry.go]: I don't think you need `found` here. You could just check in L141 if `layer == nil`, or?

File: pkg/utils/oci/helmregistry_test.go [hunk 16]
```diff
@@ -0,0 +1,160 @@
...
+	gardencorev1 "github.com/gardener/gardener/pkg/apis/core/v1"
+)
+
+var _ = Describe("helmregistry", func() {
+	var (
+		hr  *HelmRegistry
+		rc  *recordingCache
+		ctx context.Context
+	)
+	BeforeEach(func() {
```
Comment (rfranzke) [inline pkg/utils/oci/helmregistry_test.go:28]: Nit: readability

```suggestion

	BeforeEach(func() {
```

File: pkg/utils/oci/helmregistry_test.go [hunk 17]
```diff
@@ -0,0 +1,160 @@
...
+	})
+
+	It("should return error if the repository does not exist", func() {
+		_, err := hr.Pull(ctx, &gardencorev1.OCIRepository{
+			Repository: registryAddress + "/charts/examplexxx",
+			Tag:        "0.1.0",
+		})
+		Expect(err).To(MatchError(ContainSubstring("failed to pull artifact")))
+	})
+	It("should return error if the digest does not exist", func() {
```
Comment (rfranzke) [inline pkg/utils/oci/helmregistry_test.go:42]: prepending newline for readability

File: pkg/utils/oci/helmregistry_test.go [hunk 18]
```diff
@@ -0,0 +1,160 @@
...
+
+	It("should pull the chart by tag", func() {
+		out, err := hr.Pull(ctx, &gardencorev1.OCIRepository{
+			Repository: registryAddress + "/charts/example",
+			Tag:        "0.1.0",
+		})
+		Expect(err).NotTo(HaveOccurred())
+		Expect(out).NotTo(BeEmpty())
+	})
+	It("should pull the chart by digest", func() {
```
Comment (rfranzke) [inline pkg/utils/oci/helmregistry_test.go:59]: prepending newline for readability

File: pkg/utils/oci/helmregistry_test.go [hunk 19]
```diff
@@ -0,0 +1,160 @@
...
+		out, err := hr.Pull(ctx, &gardencorev1.OCIRepository{
+			Repository: registryAddress + "/charts/example",
+			Digest:     exampleChartDigest,
+		})
+		Expect(err).NotTo(HaveOccurred())
+		Expect(out).NotTo(BeEmpty())
+	})
+	It("should pull the chart with ref", func() {
```
Comment (rfranzke) [inline pkg/utils/oci/helmregistry_test.go:68]: prepending newline for readability

File: pkg/utils/oci/helmregistry_test.go [hunk 20]
```diff
@@ -0,0 +1,160 @@
...
+		}
+		_, err := hr.Pull(ctx, oci)
+		Expect(err).NotTo(HaveOccurred())
+		Expect(rc.cacheHits).To(Equal(0))
+
+		_, err = hr.Pull(ctx, oci)
+		Expect(err).NotTo(HaveOccurred())
+		Expect(rc.cacheHits).To(Equal(1))
+	})
+	It("should use the cache no matter if tag or digest is used", func() {
```
Comment (rfranzke) [inline pkg/utils/oci/helmregistry_test.go:89]: prepending newline for readability

File: pkg/utils/oci/helmregistry_test.go [hunk 21]
```diff
@@ -0,0 +1,160 @@
...
+			name.MustParseReference("garden.local.gardener.cloud:5001/foo:1.0.0", name.Insecure),
+		),
+	)
+})
+
+func mustNewTag(s string) name.Tag {
+	t, err := name.NewTag(s)
+	if err != nil {
+		panic(err)
+	}
```
Comment (rfranzke) [inline pkg/utils/oci/helmregistry_test.go]: Nit: You could replace this with `utilruntime.Must(err)`

File: pkg/utils/oci/helmregistry_test.go [hunk 22]
```diff
@@ -0,0 +1,160 @@
...
+	return t
+}
+
+func mustNewDigest(s string) name.Digest {
+	t, err := name.NewDigest(s)
+	if err != nil {
+		panic(err)
+	}
```
Comment (rfranzke) [inline pkg/utils/oci/helmregistry_test.go]: same here

File: pkg/utils/oci/oci_suite_test.go [hunk 23]
```diff
@@ -0,0 +1,76 @@
...
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+	helmRegistry "helm.sh/helm/v3/pkg/registry"
+
+	"github.com/gardener/gardener/pkg/utils"
+)
+
+func TestUtils(t *testing.T) {
+	RegisterFailHandler(Fail)
+	RunSpecs(t, "OCI Suite")
```
Comment (rfranzke) [inline pkg/utils/oci/oci_suite_test.go]:
```suggestion
	RunSpecs(t, "Utils OCI Suite")
```

File: pkg/utils/oci/oci_suite_test.go [hunk 24]
```diff
@@ -0,0 +1,76 @@
...
+	"github.com/distribution/distribution/v3/configuration"
+	"github.com/distribution/distribution/v3/registry"
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+	helmRegistry "helm.sh/helm/v3/pkg/registry"
+
+	"github.com/gardener/gardener/pkg/utils"
+)
+
+func TestUtils(t *testing.T) {
```
Comment (rfranzke) [inline pkg/utils/oci/oci_suite_test.go]:
```suggestion
func TestOCI(t *testing.T) {
```

File: pkg/utils/oci/oci_suite_test.go [hunk 25]
```diff
@@ -0,0 +1,76 @@
...
+	"fmt"
+	"os"
+	"testing"
+	"time"
+
+	"github.com/distribution/distribution/v3/configuration"
+	"github.com/distribution/distribution/v3/registry"
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+	helmRegistry "helm.sh/helm/v3/pkg/registry"
```
Comment (rfranzke) [inline pkg/utils/oci/oci_suite_test.go]: typically, we don't use camelCase in import aliases 👀 

```suggestion
	helmregistry "helm.sh/helm/v3/pkg/registry"
```

File: test/integration/gardenlet/controllerinstallation/controllerinstallation/controllerinstallation_test.go [hunk 26]
```diff
@@ -141,6 +144,45 @@ var _ = Describe("ControllerInstallation controller tests", func() {
 		})
 	})
 
+	Context("responsible with OCI", func() {
+		BeforeEach(func() {
+			DeferCleanup(test.WithVar(&controllerinstallation.RequeueDurationWhenResourceDeletionStillPresent, 500*time.Millisecond))
+		})
+		BeforeEach(func() {
```
Comment (rfranzke) [inline test/integration/gardenlet/controllerinstallation/controllerinstallation/controllerinstallation_test.go]: prepending newline for readability

File: test/integration/gardenlet/controllerinstallation/controllerinstallation/controllerinstallation_test.go [hunk 27]
```diff
@@ -141,6 +144,45 @@ var _ = Describe("ControllerInstallation controller tests", func() {
...
+			oci := &gardencorev1.OCIRepository{
+				Repository: "test",
+				Tag:        "0.1.0",
+			}
+			controllerDeployment.Helm = &gardencorev1.HelmControllerDeployment{
+				OCIRepository: oci,
+			}
+			fakeRegistry.AddArtifact(oci, rawChart)
+		})
+		It("should deploy the chart", func() {
```
Comment (rfranzke) [inline test/integration/gardenlet/controllerinstallation/controllerinstallation/controllerinstallation_test.go:162]: prepending newline for readability

File: docs/extensions/controllerregistration.md [hunk 28]
```diff
@@ -87,9 +90,39 @@ helm:
...
+  ociRepository:
+    # when specifying both tag and digest, the tag is ignored.
+    repository: registry.example.com
+    tag: 1.0.0
+    digest: sha256:abc
+```
 
+Gardenlet caches the downloaded chart in memory. It is recommended to always specify a digest, because if it is not specified, gardenlet needs to fetch the manifest in every reconciliation to compare the digest with the local cache.
+
+No matter, where the chart originates from, gardenlet deploys it with the provided static configuration (`.helm.values`).
```
Comment (rfranzke) [inline docs/extensions/controllerregistration.md]:
```suggestion
No matter where the chart originates from, gardenlet deploys it with the provided static configuration (`.helm.values`).
```

**Review by rfranzke — CHANGES_REQUESTED** [review]
Love it. Very nice PR, I'm very much looking forward to this great enhancement. Thanks a lot for your work ❤️ 

Question apart from the inline comments: Should we deprecate the `rawChart` field and plan its removal end of 2024 or so?

File: pkg/apis/core/v1/types_controllerdeployment.go (hunk 2 above)
Comment (maboehm) [reply pkg/apis/core/v1/types_controllerdeployment.go]: then I consequently have to make them all `*string`, right? As you could have just `ref` or `repository+tag` or `repository+digest`

File: pkg/apis/core/validation/controllerdeployment.go (hunk 8 above)
Comment (maboehm) [reply pkg/apis/core/validation/controllerdeployment.go:66]: that's what this does, it's XOR. It's covered by a testcase as well ("should not allow both rawChart and OCIRepository")
I'll add a comment

File: pkg/apis/core/validation/controllerdeployment_test.go (hunk 10 above)
Comment (maboehm) [reply pkg/apis/core/validation/controllerdeployment_test.go]: there are still contexts for `helm type with rawChart` and `helm type with ociRepository` and I wanted to avoid deeply nesting stuff, but I guess its not too deep yet compared to other tests

File: pkg/apis/core/v1/types_controllerdeployment.go (hunk 2 above)
Comment (rfranzke) [reply pkg/apis/core/v1/types_controllerdeployment.go]: Yes, correct 👍🏻 Validation must ensure that one of the both options is provided. :)

File: pkg/apis/core/validation/controllerdeployment.go (hunk 8 above)
Comment (rfranzke) [reply pkg/apis/core/validation/controllerdeployment.go:66]: Aha, got it 😄 I didn't see the `==` in between. Thanks!

File: pkg/utils/oci/helmregistry.go (hunk 13 above)
Comment (maboehm) [reply pkg/utils/oci/helmregistry.go]: my thougt was - this is just for caching. If the registry itself has errors in general, it will probably show up somewhere else...
But lets be diligent, I'll make `cacheKeyFromRef` return errors as well.

File: pkg/apis/core/v1/types_controllerdeployment.go (hunk 2 above)
Comment (maboehm) [reply pkg/apis/core/v1/types_controllerdeployment.go]: done. PTAL

File: pkg/utils/oci/helmregistry_test.go [hunk 29]
```diff
@@ -0,0 +1,167 @@
...
+			name.MustParseReference("garden.local.gardener.cloud:5001/foo:1.0.0", name.Insecure),
+		),
+	)
+})
+
+func mustNewTag(s string) name.Tag {
+	t, err := name.NewTag(s)
+	if err != nil {
+		utilruntime.Must(err)
+	}
```
Comment (rfranzke) [inline pkg/utils/oci/helmregistry_test.go]:
```suggestion
	utilruntime.Must(err)
```

File: pkg/utils/oci/helmregistry_test.go [hunk 30]
```diff
@@ -0,0 +1,167 @@
...
+	return t
+}
+
+func mustNewDigest(s string) name.Digest {
+	t, err := name.NewDigest(s)
+	if err != nil {
+		utilruntime.Must(err)
+	}
```
Comment (rfranzke) [inline pkg/utils/oci/helmregistry_test.go]:
```suggestion
	utilruntime.Must(err)
```

File: skaffold.yaml [hunk 31]
```diff
@@ -591,7 +591,7 @@ build:
           gitCommit:
             variant: AbbrevCommitSha
   artifacts:
-    - image: europe-docker.pkg.dev/gardener-project/releases/gardener/extensions/provider-local
+    - image: gardener-extension-provider-local
```
Comment (rfranzke) [inline skaffold.yaml]: In https://github.com/gardener/gardener/pull/9782, we just decided to rename them following this pattern: `local-skaffold/<name>`

```suggestion
    - image: local-skaffold/gardener-extension-provider-local
```

File: skaffold.yaml [hunk 32]
```diff
@@ -754,11 +754,22 @@ build:
         ldflags:
           - '{{.LD_FLAGS}}'
         main: ./cmd/gardener-extension-provider-local
+    - image: gardener-extension-provider-local/charts/extension
```
Comment (rfranzke) [inline skaffold.yaml]:
```suggestion
    - image: local-skaffold/gardener-extension-provider-local/charts/extension
```

**Review by rfranzke — CHANGES_REQUESTED** [review]
Thanks @maboehm 

Could you also adapt the [example file](https://github.com/gardener/gardener/blob/master/example/25-controllerdeployment.yaml)?

And WDYT about the question from https://github.com/gardener/gardener/pull/9823#pullrequestreview-2073138921:

> Question apart from the inline comments: Should we deprecate the `rawChart` field and plan its removal end of 2024 or so?

**Comment by maboehm** [comment]
> Should we deprecate the rawChart field and plan its removal end of 2024 or so?

I personally don't think we should remove this, even if I prefer OCI over base64. Using base64 makes it self contained, no need for a registry to push to.

**Comment by rfranzke** [comment]
> > Should we deprecate the rawChart field and plan its removal end of 2024 or so?
> 
> I personally don't think we should remove this, even if I prefer OCI over base64. Using base64 makes it self contained, no need for a registry to push to.

Fair enough, then let's keep it for now as is 👍🏻

**Comment by rfranzke** [comment]
@maboehm pls rebase

File: pkg/apis/core/validation/controllerdeployment.go [hunk 33]
```diff
@@ -58,9 +59,48 @@ func ValidateControllerDeploymentUpdate(new, _ *core.ControllerDeployment) field
...
+
+func validateOCIRepository(oci *core.OCIRepository, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+
+	if oci == nil {
+		return allErrs
+	}
+
+	if oci.Ref == "" && oci.Repository == "" {
+		allErrs = append(allErrs, field.Required(fldPath, "must provide either ref or repository"))
```
Comment (timuthy) [inline pkg/apis/core/validation/controllerdeployment.go:81]:
```suggestion
		allErrs = append(allErrs, field.Required(fldPath, "must provide either ref or repository"))
		return allErrs
```

Otherwise, `oci.Repository` are showed in addition which can be confusing.

File: pkg/gardenlet/controller/controllerinstallation/controllerinstallation/reconciler.go [hunk 34]
```diff
@@ -242,7 +244,17 @@ func (r *Reconciler) reconcile(
 		},
 	}
 
-	release, err := r.SeedClientSet.ChartRenderer().RenderArchive(helmDeployment.RawChart, controllerRegistration.Name, namespace.Name, utils.MergeMaps(helmValues, gardenerValues))
+	archive := helmDeployment.RawChart
+	if len(archive) == 0 {
```
Comment (timuthy) [inline pkg/gardenlet/controller/controllerinstallation/controllerinstallation/reconciler.go:248]: Can you add a doc string to the `ControllerDeployment` API or to the documentation in general, explaining that the `RawChart` takes precedence over `OCIRepository`?

**Review by timuthy — CHANGES_REQUESTED** [review]
Very nice PR, thanks for the contribution 🙂

File: pkg/gardenlet/controller/controllerinstallation/controllerinstallation/reconciler.go (hunk 34 above)
Comment (maboehm) [reply pkg/gardenlet/controller/controllerinstallation/controllerinstallation/reconciler.go:248]: through validation it is ensured, that it is always etiher `rawChart` or `OCIRepository`, so there is nothing to document in my opinion.

**Review by timuthy — COMMENTED** [review]
/lgtm

**Comment by maboehm** [comment]
Im looking into the tests, I think I found a solution, testing locally right now...
